### ISSUE TITLE
[Feature request] Raw pseudo classes

### ISSUE THREAD

**Issue by scripthunter7** [csstree/csstree#206]
@lahmatiy
Hi!

The idea is to add a "`rawPseudos`" option to the parser options. One of the advantages of this library is that it handles custom pseudo classes, e.g. `:contains()` selects those elements that contain the text passed to contains as a parameter:
```css
div:contains(text)
```

However, there are 1-2 exceptional cases when parsing fails, for example, the following selector fails because of the `'` character:
```css
div:contains(i'd like)
```

But if I could specify in the parsing settings that the parameters of `contains` are always read as `Raw`, then this would not be a problem:
```js
parse(`div:contains(i'd like)`, {
  rawPseudos: ['contains'],
  onParseError: (...) => { ... }
}
```

More information about Extended CSS: https://github.com/AdguardTeam/ExtendedCss

**Comment by scripthunter7** [comment]
My question was wrong. By default, it would read the pseudo content as Raw. However, things get stuck at the tokenizer level, as the library considers the input as an unfinished string:

| Token type | Start index | End index | Source part |
| --- | --- | --- | --- |
| ident-token | 0 | 6 | script
| colon-token | 6 | 7 | :
| function-token | 7 | 17 | cont2ains(
| ident-token | 17 | 18 | i
| string-token | 18 | 26 | 'd like)

I modified the tokenizer to make it work. When all ExtendedCss elements are ready, I will open a PR.

**Comment by lahmatiy** [comment]
Hi @scripthunter7,
Any changes to the tokenizer are not possible because all the rules are strictly defined in the [CSS syntax module](https://www.w3.org/TR/css-syntax-3/) and deviating from them will result in unexpected behaviour. The `url()` syntax without quotes is a single exception and considered as a legacy that should be avoided to use. Any extensions to CSS syntax must be based on CSS token's defined in CSS syntax module. For instance, to avoid issues you have you should use `div:contains("i'd like")` instead of `div:contains(i'd like)`.
You also can extend csstree's default dictionary by using `fork()` method:
```js
const mySyntax = fork({
    contains: {
        parse() {
            return this.Raw(...);
        }
    }
});

const ast = mySyntax.parse(`div:contains("i'd like"){ ... }`);
const css = mySyntax.generate(ast); // no changes into a generator in this example, but if any it will be needed
```

**Comment by scripthunter7** [comment]
@lahmatiy Thanks for the fork idea! This way I can create an independent adblock syntax. Unfortunately, the tokenize issue cannot be solved, so I have no better idea than a preprocessor before the tokenizer, which e.g. puts an escaping before the quote mark: `:contains(a'b)` -> `:contains(a\'b)`

**Comment by scripthunter7** [comment]
@lahmatiy I made a fork based on your description. If you have time to check it out, I'd be grateful for your feedback :) I hope it's not a problem that I'm using the logo and the bundle includes the entire library.

Anyway, I started writing the built-in dts files, supplemented with JSDocs. When I finish it to an acceptable level, I will send a PR. In this way, I think now I have a relatively good understanding of how the library works :)

Here is my fork:
https://github.com/scripthunter7/ecsstree
